File: 15.3-sum.java
/*
 * @lc app=leetcode id=15 lang=java
 *
 * [15] 3Sum
 */

// @lc code=start

import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        Arrays.sort(nums);

        List<List<Integer>> res = new ArrayList();

        for (int i = 0; i < nums.length - 2; i++) {

            int target = -nums[i];
            int left = i + 1;
            int right = nums.length - 1;

            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            while (left < right) {
                if (nums[left] + nums[right] < target) {
                    left++;
                } else if (nums[left] + nums[right] > target) {
                    right--;
                } else {
                    List<Integer> tempResult = List.of(nums[i], nums[left], nums[right]);
                    res.add(tempResult);

                    while (left < right && nums[left] == tempResult.get(1)) {
                        left++;
                    }

                    while (left < right && nums[right] == tempResult.get(2)) {
                        right--;
                    }
                }
            }
        }

        return res;
    }
}
// @lc code=end

File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        res = []
        i = 0

        while i < len(nums):
            
            target = -nums[i]
            left = i + 1
            right = len(nums) - 1

            while left < right:
                if nums[left] + nums[right] < target:
                    left += 1
                elif nums[left] + nums[right] > target:
                    right -= 1
                else:
                    temp_res = [nums[i], nums[left], nums[right]]
                    res.append(temp_res)

                    while left < right and nums[left] == temp_res[1]:
                        left += 1
                    while left < right and nums[right] == temp_res[2]:
                        right -= 1
            
            while i < len(nums) - 1 and nums[i] == nums[i + 1]:
                i += 1
            
            i += 1
        
        return res

        
# @lc code=end

